modules/persistence/src/services/metadata/ToC/index.ts

```diff
@@ -0,0 +1,77 @@
+import { SharedMetadata, AssociatedProduct } from '../associated_products';
+
+export interface ToC {
+  title: string;
+  slug: string;
+  children: ToC[];
+  options?: {
+    [key: string]: any;
+  };
+  [key: string]: any;
+}
+
+type project = string;
+type branchName = string;
+type branch = { [key: branchName]: ToC };
+
+export interface ToCInsertions {
+  [key: project]: branch;
+}
+
+export interface TocOrderInsertions {
+  [key: string]: {
+    [key: string]: any[];
+  };
+}
+
+const isInsertionCandidateNode = (node: ToC, associated_products: AssociatedProduct[] = []) => {
+  const nodeInProducts = node.options?.project && associated_products.find((p) => p.name === node.options?.project);
+  const nodeHasNoChildren = !node.children || node.children.length === 0;
+  return nodeHasNoChildren && nodeInProducts;
+};
+
+const mergeNode = (node: ToC, tocs: ToCInsertions) => {
+  // Options might be undefined, so safely cast to {} if nullish
+  node.options = node.options ?? {};
+
+  const project = tocs[node?.options?.project];
+  const branches = Object.keys(project);
+  node.options.versions = branches;
+  node.children = branches.map((branch) => {
+    const child = project[branch];
+    const options = {
+      ...child[0].options,
```

Comment:
we are using index numbers for an object here (child[0] is throwing an error)

```
tocs is in form of tocInsertions:
{
  project: {
    branchName: ToC
  }
}
```

so a child is a ToC is an object, not a list. not sure exactly what the intent is here